pkg/ebpf/tracee.go

```diff
@@ -277,6 +277,15 @@ func New(cfg config.Config) (*Tracee, error) {
 		t.handleEventsDependencies(id, state)
 	}
 
+	// The GKE kernel lacks CONFIG_KALLSYMS_ALL enabled, so to silence the
+	// error of missing kernel symbols, the PrintSyscallTable event is being
+	// disabled when HookedSyscalls is not requested by the user. This is a
+	// temporary solution until the sys_call_table address can be retrieved
+	// from the kernel in a different way than '/proc/kallsyms'. See #3397
+	if !t.policyManager.IsEventEnabled(events.HookedSyscalls) {
```

Comment:
`handleEventsDependencies()` sets `Submit` for both `PrintSyscallTable` and `HookedSyscalls` even if the latter is not required by any policy (`./dist/tracee -s comm=uname`):

```json
{"level":"info","ts":1696876744.1140366,"msg":"BEFORE handleEventsDependencies() - hooked_syscalls","enabled":false}
{"level":"info","ts":1696876744.1140738,"msg":"BEFORE handleEventsDependencies() - print_syscall_table","enabled":false}
{"level":"info","ts":1696876744.1141143,"msg":"depEventId","depEventId":755}
{"level":"info","ts":1696876744.1141388,"msg":"handleEventsDependencies() - givenEvtState","Submit":1}
{"level":"info","ts":1696876744.1141553,"msg":"depEventId","depEventId":2017}
{"level":"info","ts":1696876744.1141682,"msg":"handleEventsDependencies() - givenEvtState","Submit":1}
{"level":"info","ts":1696876744.114186,"msg":"AFTER handleEventsDependencies() - hooked_syscalls","enabled":true}
{"level":"info","ts":1696876744.1141973,"msg":"AFTER handleEventsDependencies() - print_syscall_table","enabled":true}
```

The unique way to know, at this stage, if `HookedSyscalls` is required by the user, is checking the policies.

So, we should disable both events, not only `PrintSyscallTable`. I'm changing that.

Regarding not using policyManager, if so, we would have to do this instead:

```go
	hookedSyscallsRequired := false
	for p := range t.config.Policies.Map() {
		if _, ok := p.EventsToTrace[events.HookedSyscalls]; !ok {
			continue
		}
		hookedSyscallsRequired = true
		break
	}
	if !hookedSyscallsRequired {
		delete(t.eventsState, events.HookedSyscalls)
		delete(t.eventsState, events.PrintSyscallTable)
	}
```

That's why I opted to:

```go
	if !t.policyManager.IsEventEnabled(events.HookedSyscalls) {
		delete(t.eventsState, events.HookedSyscalls)
		delete(t.eventsState, events.PrintSyscallTable)
	}
```

Reply:
I didnt mean both events (which you could be right) but in both "places" (the t.eventsState AND the policy manager). About the dependency, maybe you should disable the events depending on the broken hook and later run handle dependencies.

Reply:
> I didnt mean both events (which you could be right) but in both "places" (the t.eventsState AND the policy manager).

Ok, I got it now. For that, it's required only to remove it from eventState since this removal is only when the event is not required in any policy. If it's we continue emitting the error if we can't get the sys_call_table address.

> About the dependency, maybe you should disable the events depending on the broken hook and later run handle dependencies.

Agreed and implemented a inner function to iterate and disable it. Meanwhile doing that I've detected the real requirement (default) of HookedSyscalls: `6028 syscall_hooking` - a signature event.

Running with `--signatures-dir=/tmp` silences the error.

Reply:
Based on this, I don't really agree with doing this workaround to disable a sig event that might fail in a specific environment. If the environment does not meet the requirements for that event, it must be cancelled, right? And this is already being done.

So what we can really discuss is whether we should continue considering this failure as an error (my choice) or change it to a warning just so as not to scare the GKE user. If we decide not to log an error, I suggest disabling the `syscall_hooking` signature by default.

Reply:
Let me see if I got this right.

1. We have our default signatures enabled (tracee-policies.yaml)
2. It has the event "syscall_hooking" enabled by default.
3. "syscall_hooking" (HookedSyscalls) depend on PrintSyscallTable event.
4. PrintSyscallTable depends on ksym "sys_call_table"
5. Some GKE kernels don't have that symbol in kallsyms file.
6. Tracee fails to initialize entirely because of that.

Despite any alternatives (of finding the symbol address in a way that does not require reading kallsyms) the intent here is to load tracee disabling the policies (or just rules) that relies in PrintSyscallTable, for example.

Your solution was an attempt of deleting the events from eventsState. I confess I haven't read entirely the changes in the policies code (recent ones) but IMO:

1. We should disable the RULE of any POLICY that uses PrintSyscallTable (the event directly)
2. We should disable any RULE that depends on the PrintSyscallTable (the event directly)
3. We should disable signatures using PrintSyscallTable or anything depending on it.
4. And, of course, warn the user that has been disabled.

But anything you come up with should be "generic" enough to work for other events as well (there are other events depending on ksyms that could be nonexistent as well).

Reply:
> Tracee fails to initialize entirely because of that.

Since #3495 tracee continues running even when failing to attach probes.

Well, my current proposal takes care of disabling events which can't be traced beforehand - `prepareEventsToTrace` -, based on the `EventsToDisable` hardcoded map.

Reply:
I left a comment there, but it should be put here as well:

https://github.com/aquasecurity/tracee/pull/3495/files#r1353963525

Reply:
Thanks for your feedback @rafaeldtinoco. I just made changes accordingly and raise them.